File: src/simcore/quantities.py
"""Conserved quantities, stock kinds, and the canonical-unit table.

This is part of the frozen Phase-0 API (see ``docs/plans/phase-0-engine-skeleton``).
Everything here is plain stdlib data — no third-party imports.
"""

from enum import Enum

from simcore.ids import UnitLabel


class Quantity(Enum):
    """A conserved quantity tracked by the ledger.

    The conservation gate balances each member independently every step. Add a
    member only when a real stock and the matching flow stoichiometry exist for
    it — every per-quantity loop and the canonical-unit table below must cover
    each member, so an unused quantity is dead weight.

    ``PHOSPHORUS`` is intentionally *reserved as a comment*, not a member: the
    design anticipates it, but it has no Phase-0 stock or unit. (Matches the
    frozen API, which lists it as ``# PHOSPHORUS reserved``.)
    """

    CARBON = "carbon"
    WATER = "water"
    NITROGEN = "nitrogen"
    OXYGEN = "oxygen"
    ENERGY = "energy"
    # PHOSPHORUS reserved — add when a phosphorus stock + flow actually exist.


class StockKind(Enum):
    """How a stock behaves under arbitration and extinction.

    - ``POOL``: a resource pool (e.g. atmospheric carbon, water). Never
      zeroed-with-loss; arbitration may throttle draws against it.
    - ``POPULATION``: absorbing-eligible biomass/population. May go extinct —
      below its ``extinction_threshold`` it snaps to 0 and the residual is routed
      to the numerical-loss boundary sink (decision #6).
    - ``BOUNDARY``: an "outside" reservoir. Its per-step delta *is* a ledger
      Input/Output (decision #13). May be flagged ``unclamped`` (e.g. solar) so
      arbitration's min-scaling never throttles it.
    """

    POOL = "pool"
    POPULATION = "population"
    BOUNDARY = "boundary"


# Canonical-unit table — the single shared source of truth (decision #9). The
# core carries only these *labels*; the outer ``config`` loader validates and
# converts incoming params against them with pint.
#
# RESOLVED (Phase-1 Step 1; was Phase-0 PROVISIONAL). Conservation arithmetic only
# requires one consistent canonical unit per quantity, but the science picks are
# now fixed:
#   - CARBON = mol, ENERGY = J  → GOLDEN-LOCKED: the committed demo regression
#     goldens (``tests/regression/golden/demo_*``) carry these labels, so changing
#     either forces regenerating those goldens. ``mol C`` / ``J`` are adequate.
#   - WATER = kg, NITROGEN = kg → mass basis. kg H2O is the Penman-Monteith
#     convention (transpiration mm/day = kg m^-2 day^-1); kg N is unambiguous
#     element mass (WOFOST-native, unlike species-ambiguous "mol N"). (Flipping
#     WATER from the Phase-0 ``mol`` placeholder regenerated the one golden that
#     carries a water stock, ``tests/regression/golden/state_snapshot.json``.)
#   - OXYGEN = mol → untracked in Phase 1 (no O2 stock/flow); molar keeps the
#     gas-phase species consistent for the deferred Phase-2 stoichiometry.
# Per-organ *biomass* is conventionally kg dry matter; our currency is CARBON in
# mol, so kg-DM <-> mol-C is a boundary conversion in the biosphere loader (carbon
# fraction + molar mass), NOT a second canonical unit for carbon.
CANONICAL_UNIT: dict[Quantity, UnitLabel] = {
    Quantity.CARBON: UnitLabel("mol"),
    Quantity.WATER: UnitLabel("kg"),
    Quantity.NITROGEN: UnitLabel("kg"),
    Quantity.OXYGEN: UnitLabel("mol"),
    Quantity.ENERGY: UnitLabel("J"),
}


def canonical_unit(quantity: Quantity) -> UnitLabel:
    """Return the canonical-unit label for ``quantity``.

    Raises ``KeyError`` if a quantity has no table entry — which would be a bug,
    since every ``Quantity`` member must be covered (a test asserts totality).
    """
    return CANONICAL_UNIT[quantity]


# --- balance contract (shared by the step-3 flow check + step-8 ledger) ------
# The quantities whose per-flow / per-step balance is ASSERTED. ENERGY is exempt
# (decision #8: energy *closure* is Phase 5/6; here energy is a diagnostic only).
# This is the single source of truth for "what we assert" — co-located with the
# tolerances below ("how tightly") so the step-3 flow-balance helper and the
# step-8 conservation gate cannot drift apart.
ASSERTED_QUANTITIES: frozenset[Quantity] = frozenset(Quantity) - {Quantity.ENERGY}

# The balance test is ``abs(residual) <= BALANCE_ATOL + BALANCE_RTOL * scale``,
# where ``scale`` is the transfer magnitude (max ``|leg.amount|`` for the
# quantity). The relative term keeps the check meaningful over the varied
# magnitudes the step-8 every-step gate sees. Phase-0 defaults, verified adequate
# against the demo's real magnitudes (step-11 conservation-tol carry-forward:
# ``BALANCE_ATOL`` floors ``tol`` above the stored-rounding floor — see
# ``conservation.assert_conserved``). The *form* matters, not these exact values.
BALANCE_ATOL: float = 1e-9
BALANCE_RTOL: float = 1e-9

File: rust/crates/simcore/src/quantities.rs
//! Conserved quantities, stock kinds, and the canonical-unit table — the Rust port
//! of `simcore.quantities`.
//!
//! [`Quantity`] members carry two string projections, both mirrored from the Python
//! `Enum`: [`Quantity::value`] (lowercase — `"carbon"`, used in loss-sink ids,
//! snapshot JSON, and composition keys) and [`Quantity::name`] (uppercase —
//! `"CARBON"`, used in the by-name sort order and error messages). The port replicates
//! both so canonical ordering and diagnostics match byte-for-byte.
//!
//! `ENERGY` is in the asserted conserved set (it joined in Phase 5 — the Power
//! domain); the balance gate asserts every member independently.

use crate::ids::UnitLabel;

/// A conserved quantity tracked by the ledger. `PHOSPHORUS` is reserved in the
/// Python source as a comment (no stock/unit yet); it is intentionally absent here.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Quantity {
    Carbon,
    Water,
    Nitrogen,
    Oxygen,
    Energy,
}

impl Quantity {
    /// The lowercase canonical value (`"carbon"`) — the Python `Enum` *value*, used
    /// in loss-sink ids, snapshot JSON, and composition keys.
    pub fn value(&self) -> &'static str {
        match self {
            Quantity::Carbon => "carbon",
            Quantity::Water => "water",
            Quantity::Nitrogen => "nitrogen",
            Quantity::Oxygen => "oxygen",
            Quantity::Energy => "energy",
        }
    }

    /// The uppercase member name (`"CARBON"`) — the Python `Enum` *name*, used for
    /// the by-name canonical sort and in error messages.
    pub fn name(&self) -> &'static str {
        match self {
            Quantity::Carbon => "CARBON",
            Quantity::Water => "WATER",
            Quantity::Nitrogen => "NITROGEN",
            Quantity::Oxygen => "OXYGEN",
            Quantity::Energy => "ENERGY",
        }
    }

    /// The canonical-unit label for this quantity (single source of truth, #9).
    /// Total by construction — every member has an entry (Python raises `KeyError`
    /// on a gap; here the match is exhaustive, so a gap is a compile error).
    pub fn canonical_unit(&self) -> UnitLabel {
        match self {
            Quantity::Carbon => "mol",
            Quantity::Water => "kg",
            Quantity::Nitrogen => "kg",
            Quantity::Oxygen => "mol",
            Quantity::Energy => "J",
        }
        .to_string()
    }
}

/// Every `Quantity` in `.name`-sorted order — the canonical iteration order the
/// balance gate uses so the first reported failure is deterministic. Sorting the
/// uppercase names `CARBON < ENERGY < NITROGEN < OXYGEN < WATER` gives this fixed
/// list (mirrors Python's `sorted(ASSERTED_QUANTITIES, key=lambda q: q.name)`).
pub const ASSERTED_QUANTITIES: [Quantity; 5] = [
    Quantity::Carbon,
    Quantity::Energy,
    Quantity::Nitrogen,
    Quantity::Oxygen,
    Quantity::Water,
];

/// How a stock behaves under arbitration and extinction.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StockKind {
    /// A resource pool — never zeroed-with-loss; arbitration may throttle draws.
    Pool,
    /// Absorbing-eligible biomass/population — may go extinct (snap to 0, route the
    /// residual to the loss-sink).
    Population,
    /// An "outside" reservoir — its per-step delta *is* a ledger Input/Output.
    Boundary,
}

impl StockKind {
    /// The lowercase canonical value (`"pool"`) — the Python `Enum` value, used in
    /// snapshot JSON.
    pub fn value(&self) -> &'static str {
        match self {
            StockKind::Pool => "pool",
            StockKind::Population => "population",
            StockKind::Boundary => "boundary",
        }
    }

    /// The uppercase member name (`"POOL"`) — used in error messages.
    pub fn name(&self) -> &'static str {
        match self {
            StockKind::Pool => "POOL",
            StockKind::Population => "POPULATION",
            StockKind::Boundary => "BOUNDARY",
        }
    }
}

/// The balance tolerance: `abs(residual) <= BALANCE_ATOL + BALANCE_RTOL * scale`.
/// Same Phase-0 values as the Python contract (`simcore.quantities`).
pub const BALANCE_ATOL: f64 = 1e-9;
/// The relative term of the balance tolerance (see [`BALANCE_ATOL`]).
pub const BALANCE_RTOL: f64 = 1e-9;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_units_match_python_table() {
        assert_eq!(Quantity::Carbon.canonical_unit(), "mol");
        assert_eq!(Quantity::Water.canonical_unit(), "kg");
        assert_eq!(Quantity::Nitrogen.canonical_unit(), "kg");
        assert_eq!(Quantity::Oxygen.canonical_unit(), "mol");
        assert_eq!(Quantity::Energy.canonical_unit(), "J");
    }

    #[test]
    fn asserted_order_is_name_sorted() {
        // The exact order Python's `sorted(..., key=q.name)` yields.
        let names: Vec<&str> = ASSERTED_QUANTITIES.iter().map(|q| q.name()).collect();
        assert_eq!(names, ["CARBON", "ENERGY", "NITROGEN", "OXYGEN", "WATER"]);
        let mut sorted = names.clone();
        sorted.sort_unstable();
        assert_eq!(names, sorted);
    }
}
